presto-native-execution/presto_cpp/main/TaskManager.cpp

```diff
@@ -1098,16 +1098,41 @@ bool TaskManager::getStuckOpCalls(
     std::vector<velox::exec::Task::OpCallInfo>& stuckOpCalls) const {
   const auto thresholdDurationMs =
       SystemConfig::instance()->driverStuckOperatorThresholdMs();
+  const auto thresholdCancelMs =
+      SystemConfig::instance()
+          ->driverCancelTasksWithStuckOperatorsThresholdMs();
+  stuckOpCalls.clear();
+
   const std::chrono::milliseconds lockTimeoutMs(thresholdDurationMs);
   auto taskMap = taskMap_.rlock(lockTimeoutMs);
   if (!taskMap) {
     return false;
   }
-  for (const auto& [_, prestoTask] : *taskMap) {
-    if (prestoTask->task != nullptr &&
-        !prestoTask->task->getLongRunningOpCalls(
-            lockTimeoutMs, thresholdDurationMs, stuckOpCalls)) {
-      deadlockTasks.push_back(prestoTask->task->taskId());
+
+  for (const auto& [id, prestoTask] : *taskMap) {
+    if (prestoTask->task != nullptr) {
+      const auto numPrevStuckOps = stuckOpCalls.size();
+      if (!prestoTask->task->getLongRunningOpCalls(
+              lockTimeoutMs, thresholdDurationMs, stuckOpCalls)) {
+        deadlockTasks.push_back(prestoTask->task->taskId());
+        continue;
+      }
+      // See if we need to cancel the Task - it should be running, the cancel
+      // threshold should be valid and it should have at least one stuck driver
+      // that was stuck for enough time.
+      if (numPrevStuckOps < stuckOpCalls.size() && thresholdCancelMs != 0 &&
+          prestoTask->task->isRunning()) {
+        for (auto it = stuckOpCalls.begin() + numPrevStuckOps;
+             it != stuckOpCalls.end();
+             ++it) {
+          if (it->durationMs >= thresholdCancelMs) {
+            LOG(INFO) << "Cancelling task with stuck drivers '" << id << "'.";
+            prestoTask->task->requestCancel();
```

Comment:
Not sure if it is better to call setError instead of requestCancel() as the latter might be used by coordinator for task state control as requestAbort() is used by coordinator to abort a task execution if it fails at coordinator side.